cloudify_aws/ec2/resources/image.py

```diff
@@ -77,28 +83,87 @@ def create(self, params):
         return res
 
     def delete(self, params=None):
-        return
+        self.logger.debug('Deleting %s' % self.type_name)
+        self.logger.debug('Deregistering ImageId %s' % params.get('ImageId'))
+        self.client.deregister_image(**params)
+
+    def prepare_describe_image_filter(self, params):
 
+        dry_run = params.get(DRY_RUN, False)
+        image_ids = params.get(IMAGE_IDS, [])
+        owners = params.get(OWNERS, [])
+        executable_users = params.get(EXECUTABLE_USERS, [])
+        filters = params.get(FILTERS, [])
+        if not any([dry_run, image_ids, owners, executable_users, filters]):
+            return
+        self.describe_image_filters = {
+            DRY_RUN: dry_run,
+            IMAGE_IDS: image_ids,
+            OWNERS: owners,
+            EXECUTABLE_USERS: executable_users,
+            FILTERS: filters
+        }
 
-def prepare_describe_image_filter(params, iface):
-    iface.describe_image_filters = {
-        DRY_RUN: params.get(DRY_RUN, False),
-        IMAGE_IDS: params.get(IMAGE_IDS, []),
-        OWNERS: params.get(OWNERS, []),
-        EXECUTABLE_USERS: params.get(EXECUTABLE_USERS, []),
-        FILTERS: params.get(FILTERS, [])
-    }
-    return iface
+
+def clean_params(params):
+    if isinstance(params, dict):
+        new_params = {}
+        for key, val in params.items():
+            if isinstance(val, dict) or isinstance(val, list):
+                val = clean_params(val)
+            if val:
+                new_params[key] = val
+        return new_params
+    if isinstance(params, list):
+        new_params = []
+        for val in params:
+            if isinstance(val, dict) or isinstance(val, list):
+                val = clean_params(val)
+            if val:
+                new_params.append(val)
+        return new_params
 
 
 @decorators.aws_resource(EC2Image, resource_type=RESOURCE_TYPE)
 def prepare(ctx, iface, resource_config, **_):
     """Prepares an AWS EC2 Image"""
     # Save the parameters
-    ctx.instance.runtime_properties['resource_config'] = resource_config
-    iface = \
-        prepare_describe_image_filter(
-            resource_config.copy(),
-            iface)
-    ami = iface.properties
-    utils.update_resource_id(ctx.instance, ami.get(IMAGE_ID))
+    if not iface.describe_image_filters:
+        return
+    if not ctx.instance.runtime_properties.get('use_external_resource'):
```

Comment:
Shouldn't this be `if ctx.instance.runtime_properties.get('use_external_resource')`?